src/Stack/Options/UploadParser.hs

```diff
@@ -0,0 +1,38 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+module Stack.Options.UploadParser
+  ( UploadOpts(..)
+  , UploadVariant(..)
+  , uploadOptsParser
+  ) where
+
+import           Options.Applicative
+import           Stack.Options.SDistParser (sdistOptsParser)
+import           Stack.Prelude
+import           Stack.SDist (SDistOpts(..))
+
+data UploadOpts = UploadOpts
+  { uoptsSDistOpts :: SDistOpts
+  , uoptsUploadVariant :: UploadVariant
+  -- ^ Says whether to publish the package or upload as a release candidate
+  }
+
+data UploadVariant
+  = Publishing
+  -- ^ Publish the package
+  | Candidate
+  -- ^ Create a package candidate
+
+-- | Parser for arguments to `stack upload`
+uploadOptsParser :: Parser UploadOpts
+uploadOptsParser =
+  UploadOpts
+    <$> sdistOptsParser
+    <*> uploadVariant
+  where
+    uploadVariant =
+      switch (long "candidate"
```

Comment:
How about using [`flag`](https://www.stackage.org/haddock/lts-17.7/optparse-applicative-0.15.1.0/Options-Applicative.html#v:flag) instead of `switch` and `<&>`?

Reply:
I've updated it to use `flag`. Thanks